stor/dx.py

```diff
@@ -113,96 +165,727 @@ def temp_url(self, lifetime=300, filename=None):
+        """
+        parts = self._get_parts()
+        if len(parts) == 2 and parts[1]:  # paths like ('dx://proj:file') need different logic
+            parts[0] += ':'
+            new_path = '/'.join(parts)
+            return self.parts_class(self.path_module.basename(new_path))
+        else:
+            return super(DXPath, self).name
 
     def remove(self):
-        raise NotImplementedError
+        """Removes a single object from DX platform
 
+        Raises:
+            ValueError: The path is invalid.
+        """
+        if not self.resource:
+            raise ValueError('DXPath must point to single data object to call remove')
+        file_handler = dxpy.DXFile(dxid=self.canonical_resource,
+                                   project=self.canonical_project)
+        with _propagate_dx_exceptions():
+            file_handler.remove()
+        self.clear_cached_properties()
+
+    @_propagate_dx_exceptions()
     def rmtree(self):
-        raise NotImplementedError
+        """
+        Removes a resource and all of its contents.
+        The path should point to a project or directory.
+
+        Raises:
+            NotFoundError: The path points to a nonexistent directory
+        """
+        proj_handler = dxpy.DXProject(self.canonical_project)
+        if not self.resource:
+            return proj_handler.destroy()
+        try:
+            proj_handler.remove_folder(self.resource, recurse=True)
+        except dxpy.exceptions.ResourceNotFound as e:
+            raise stor_exceptions.NotFoundError('No folders were found with the given path ({})'
+                                                .format(self), e)
+        self.clear_cached_properties()
+
+    def makedirs_p(self, mode=0o777):
+        """Make directories, including parents on DX from DX folder paths.
+
+        The resulting folders will have same access permissions as the project.
+
+        Args:
+            mode: unused, present for compatibility
+        """
+        if not self.resource:
+            if not self.exists():
+                raise ValueError('Cannot create a project via makedirs_p()')
+            return
+        proj_handler = dxpy.DXProject(self.canonical_project)
+        with _propagate_dx_exceptions():
+            proj_handler.new_folder(self.resource, parents=True)
 
     def isdir(self):
-        if not self.resource or self._is_folder():
-            return self.exists()
-        return False
+        """Determine if path is an existing directory
+
+        Returns:
+            bool: True if path is an existing folder path or project
+        """
+        if not self.resource and self.exists():  # path could be a project
+            return True
+        # or path could be a folder
+        try:
+            self.listdir(only='folders')
+            return True
+        except stor_exceptions.NotFoundError:
+            return False
 
     def isfile(self):
+        """Determine if path is an existing file
```

Comment:
Determine an object exists at the specified path

Reply:
Fixed